1_fetch/src/get_nwis_data.R

```diff
@@ -1,48 +1,102 @@
+  
+  for (download_file in download_files){                               # set start and end dates in here
+    download_nwis_site_data(download_file, parameterCd =  parameterCd, plzStart, plzStop)
     # read the downloaded data and append it to the existing data.frame
-    these_data <- read_csv(download_file, col_types = 'ccTdcc')
+    
+    these_data <- read_csv(download_file, col_types = 'ccTdcc')                  # set the return inside here
     data_out <- bind_rows(data_out, these_data)
   }
   return(data_out)
 }
 
+# optional function
+checkExsits <- function(download_files, endDate, filepath){
+  for (data in download_files) {
+    these_data <- read_csv(download_file, col_types = 'ccTdcc')    # actually I shouldn't use this
+    if (!dir.exists(data) & !data$dateTime.tail() == endDate) {
+      write.csv(data, filepath)
+    }
+    else 
+      download_nwis_site_data(data, parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01")
+  } 
+}
+
+
+# 2nd function
 nwis_site_info <- function(fileout, site_data){
   site_no <- unique(site_data$site_no)
   site_info <- dataRetrieval::readNWISsite(site_no)
   write_csv(site_info, fileout)
   return(fileout)
 }
+                   
 
-
-download_nwis_site_data <- function(filepath, parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01"){
-  
-  # filepaths look something like directory/nwis_01432160_data.csv,
+# 3rd funciton
+download_nwis_site_data <- function(filepath, parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01")
+  {  # changed this function signature too
   # remove the directory with basename() and extract the 01432160 with the regular expression match
   site_num <- basename(filepath) %>% 
-    stringr::str_extract(pattern = "(?:[0-9]+)")
+    stringr::str_extract(pattern = "(?:[0-9]+)")  # this gets the number from the csv identifier
   
-  # readNWISdata is from the dataRetrieval package
-  data_out <- readNWISdata(sites=site_num, service="iv", 
-                           parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+  if (!dir.exists(file.path("1_fetch/out", paste0('nwis_', site_nums, '_data.csv')))) # this produces warning, but the warning is only about the length of the input, since there is an input vector that is iterated through
+    {
+    # readNWISdata is from the dataRetrieval package
+    data_out <- readNWISdata(sites=site_num, service="iv", 
+                             parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+    write_csv(data_out, file = filepath)
+  }
 
   # -- simulating a failure-prone web-sevice here, do not edit --
   set.seed(Sys.time())
   if (sample(c(T,F,F,F), 1)){
-    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
+    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')    # does this means it stops at time 1000?
```

Comment:
This chunk is simulating failure (like a website that times out) by randomly sampling TRUE and FALSE (line 95) - so each time this is run you have a 25% chance of failure.